code/drv_bms/src/wattrex_driver_bms/drv_bms.py

```diff
@@ -0,0 +1,229 @@
+        """
+        while not self.__rx_chan.is_empty():
+            raw_data: DrvCanMessageC = self.__rx_chan.receive_data_unblocking()
+            self._defragment(raw_data.payload)
+        return self.__data
+
+    def __reset_raw_data(self) -> None:
+        """Reset the raw data to the initial value .
+        """
+        self.__raw_data = []
+        self.__next_idx_frag = 0
+
+
+    def _defragment(self, data) -> None:
+        '''
+        Defragment the message received from BMS with dev_id. When all fragments has 
+        been received and defragmented, the resulting message is parsed and stored
+        into the local data.
+
+        Args:
+            id (int): Identificator of the sender bms.
+            data ([type]): Data to be defragmented.
+        '''
+        # Prase ids and frags
+        frag_idx = 0xF & data[0]
+        n_frags = data[0] >> 4
+
+        # ---------------------------------------------------------------------
+
+        # Deframnet error code
+        if frag_idx == 0 and n_frags == 1 and str(data[1:]).startswith("ERROR"):
+            recv_status = _BmsStatusE(int(data[7]))
+            if recv_status == _BmsStatusE.ERROR_COMM:
+                self.__data.status = DrvBaseStatusC(DrvBaseStatusE.COMM_ERROR)
+            log.warning(f"Error fragment received on BMS: {self.__data.status}")
+
+        # Normal parse, parse normal message
+        else:
+            if self.__next_idx_frag != frag_idx:
+                log.error((f"ERROR on [0x{self.dev_id:03x}] - Not received all fragment. "
+                           f"Expected frag: {self.__next_idx_frag}. Recv frag: {frag_idx}"))
+                self.__data.status = DrvBaseStatusC(DrvBaseStatusE.COMM_ERROR)
+                self.__reset_raw_data()
+
+            else:
+                self.__raw_data.extend(data[1:])
+                # log.debug(f"Frag {self.dev_id} - [{frag_idx}/{n_frags}]")
+
+                # Prepare for new fragment reception
+                self.__next_idx_frag += 1
+                # Parse all received data
+                if frag_idx == n_frags-1:
+                    if len(self.__raw_data) % 2 != 0:
+                        log.error(f"Error on communication with BMS: Not received all fragment")
+                        self.__data.status = DrvBaseStatusC(DrvBaseStatusE.COMM_ERROR)
+                        self.__reset_raw_data()
+                    else:
+                        # Iterate self.__raw_data every 2 elements and combine them into one int
+                        new_data = []
+                        for i in range(0, len(self.__raw_data), 2):
+                            new_elem = self.__raw_data[i+1] << 8 | self.__raw_data[i]
+                            new_data.append(new_elem)
+
+                        self.__data = DrvBmsDataC(new_data)
+                        # add __str__ function
+                        log.info(f"Data received from BMS: {self.__data.__dict__}")
```

Comment:
Change it when str function is implemented

Reply:
Done, check if str function is enough or should include something more